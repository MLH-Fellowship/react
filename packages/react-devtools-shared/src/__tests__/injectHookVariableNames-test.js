/**
 * @flow
 */

import {parse} from '@babel/parser';
import {
  getHookVariableName,
  getPotentialHookDeclarationsFromAST,
  checkNodeLocation,
  isConfirmedHookDeclaration,
  getFilteredHookASTNodes,
} from 'react-devtools-shared/src/utils';

describe('injectHookVariableNamesFunction', () => {
  it('should identify variable names in destructed syntax', async done => {
    const jsxCode = `
            const Example = () => {
                const [count, setCount] = React.useState(1);
                return count;
            };
        `;

    const ast = parse(jsxCode, {
      sourceType: 'unambiguous',
      plugins: ['jsx', 'typescript'],
    });
    const hookAstNodes = getPotentialHookDeclarationsFromAST(ast);

    // Only one hook node is present in the source code
    expect(hookAstNodes).toHaveLength(1);

    const hookName = getHookVariableName(hookAstNodes[0]);
    expect(hookName).toBe('count');
    done();
  });

  it('should identify variable names in direct assignment', async done => {
    const jsxCode = `
            const Example = () => {
                const count = React.useState(1);
                return count;
            };
        `;

    const ast = parse(jsxCode, {
      sourceType: 'unambiguous',
      plugins: ['jsx', 'typescript'],
    });
    const hookAstNodes = getPotentialHookDeclarationsFromAST(ast);

    // Only one hook node is present in the source code
    expect(hookAstNodes).toHaveLength(1);

    const hookName = getHookVariableName(hookAstNodes[0]);
    expect(hookName).toBe('count');
    done();
  });

  it('should identify variable names in case of destructured assignment', async done => {
    const jsxCode = `
            const Example = () => {
                const count = React.useState(1);
                const [x, setX] = count;
                return count;
            };
        `;

    const ast = parse(jsxCode, {
      sourceType: 'unambiguous',
      plugins: ['jsx', 'typescript'],
    });
    const hookAstNodes = getPotentialHookDeclarationsFromAST(ast);
    // This line number corresponds to where the hook is present
    const lineNumber = 3;

    // Isolate the Hook AST Node
    const potentialReactHookASTNode = hookAstNodes.find(
      node =>
        checkNodeLocation(node, lineNumber) && isConfirmedHookDeclaration(node),
    );
    // Find the nodes that are associated with the React Hook found - in this case we obtain the [x, setX] line
    const nodesAssociatedWithReactHookASTNode = getFilteredHookASTNodes(
      potentialReactHookASTNode,
      hookAstNodes,
      'example-app',
      new Map(),
    );

    // Only one node should be found here
    expect(nodesAssociatedWithReactHookASTNode).toHaveLength(1);
    const relatedNode = nodesAssociatedWithReactHookASTNode[0];

    // The [x,setX] destructuring is on line 4
    expect(relatedNode.node.loc.start.line).toBe(4);

    const hookName = getHookVariableName(relatedNode);
    expect(hookName).toBe('x');
    done();
  });

  it('should identify variable names for multiple hooks in one app', async done => {
    const jsxCode = `
        const Example = () => {
            const count = React.useState(1);
            const [x, setX] = count;
            const [count1, setCount1] = useState(0);
            return count;
            };
        `;

    done();
  });
});
