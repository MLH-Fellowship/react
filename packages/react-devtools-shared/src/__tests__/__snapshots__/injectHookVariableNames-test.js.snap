// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectHookVariableNamesFunction should create hookLog injected with default variable names for no/multiple references: modified hook log 1`] = `
Array [
  Object {
    "hookSource": Object {
      "columnNumber": 42,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 3,
    },
    "hookVariableName": "countState",
    "id": 0,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": 1,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 35,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 5,
    },
    "hookVariableName": "flagState",
    "id": 1,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": false,
  },
]
`;

exports[`injectHookVariableNamesFunction should create hookLog injected with variable names for destructured assignment: modified hook log 1`] = `
Array [
  Object {
    "hookSource": Object {
      "columnNumber": 49,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 3,
    },
    "hookVariableName": "count",
    "id": 0,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": 0,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 41,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 4,
    },
    "hookVariableName": "flag",
    "id": 1,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": false,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 29,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 5,
    },
    "hookVariableName": "ref",
    "id": 2,
    "isStateEditable": false,
    "name": "Ref",
    "subHooks": Array [],
    "value": undefined,
  },
]
`;

exports[`injectHookVariableNamesFunction should create hookLog injected with variable names for inline/multi-line member expressions: modified hook log 1`] = `
Array [
  Object {
    "hookSource": Object {
      "columnNumber": 42,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 3,
    },
    "hookVariableName": "count",
    "id": 0,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": 1,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 35,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 5,
    },
    "hookVariableName": "flag",
    "id": 1,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": false,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 41,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 8,
    },
    "hookVariableName": "tick",
    "id": 2,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": 0,
  },
]
`;

exports[`injectHookVariableNamesFunction should not inject hookLog with variable names for non declarative primitive hooks: modified hook log 1`] = `
Array [
  Object {
    "hookSource": Object {
      "columnNumber": 44,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 3,
    },
    "hookVariableName": "flag",
    "id": 0,
    "isStateEditable": true,
    "name": "State",
    "subHooks": Array [],
    "value": false,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 40,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 4,
    },
    "hookVariableName": "inputRef",
    "id": 1,
    "isStateEditable": false,
    "name": "Ref",
    "subHooks": Array [],
    "value": undefined,
  },
  Object {
    "hookSource": Object {
      "columnNumber": 17,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 6,
    },
    "id": 2,
    "isStateEditable": false,
    "name": "Effect",
    "subHooks": Array [],
    "value": [Function],
  },
  Object {
    "hookSource": Object {
      "columnNumber": 17,
      "fileName": "example-app.js",
      "functionName": "Example",
      "lineNumber": 9,
    },
    "id": 3,
    "isStateEditable": false,
    "name": "LayoutEffect",
    "subHooks": Array [],
    "value": [Function],
  },
]
`;
